sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -779,6 +784,37 @@ struct _pi_queue : _pi_object {
   // asked to not transfer the ownership to SYCL RT.
   bool OwnZeCommandQueue;
 
+  // Indicates if we create an event for each command or use eventless mode for
+  // the queue. The eventless mode is used for SYCL in-order queue with
+  // discard_events property. Since we don't create L0 events for some commands
+  // we guarantee in-order semantics inside command-list by using barriers and
+  // create a special event for the last barrier of the command-list to
+  // guarantee the semantics between command-lists.
+  bool EventlessMode = false;
+
+  // It helps to skip creating a special service barrier with an event in
+  // command-list in eventless mode if we do QueueFinish or piQueueRelease
+  bool SkipLastEventInEventlessMode = false;
+
+  // Indicates if the previous command was submitted into command-list of the
+  // copy engine. it is used in eventless mode to identify that we switch
+  // command-list of copy queue to command-list of compute queue or vice versa.
+  // if the switch happened then we add a special service barrier with an event
+  // into the last command-list to maintain the order between the two
+  // command-lists.
+  bool IsPrevCopyEngine = false;
+
+  // Keeps an event of last special service barrier created to ensure in-order
+  // semantics between command-lists. The plugin is a holder of this event and
+  // SYCL RT knows nothing about it. This special event will be destroyed and
+  // returned back in pool by EventRelease in resetCommandList. this is used to
+  // compare against LastCommandEvent to ensure that event-dependency will only
+  // be added between command-lists.
+  pi_event LastEventInPrevCmdList = nullptr;
+
+  // Keeps command-list of the previous command to use in EventlessMode.
+  pi_command_list_ptr_t LastCommandList{};
```

Comment:
I also don't see why you'd need this. Why don't you create the "barrier" when closing the current command-list, remember it's event in the `_pi_queue::LastCommandEvent` and add another "barrier" at the beginning of a new command-list?

Reply:
>  Why don't you create the "barrier" when closing the current command-list, remember it's event in the _pi_queue::LastCommandEvent

We do this, but if there is a switch between сopy and compute command-list, then we must submit the barrier with an event into `LastCommandList` as it is still opened and we take the event to add to WaitList for the current command-list.
I added a comment here about it.

Reply:
Sorry, but I still don't get it. I think you should always insert "barrier" to the current command-list before closing/submitting it and not need to keep track of a previous command-list. Please illustrate in more details why you need this. With in-order queue, there can only be single open command-list, so I don't understand the comment about switching between copy and compute.

Reply:
> Thus, in eventless case, to maintain the original order of operations, when changing from one command-list to another, we must be able to submit a barrier with an event to the previous command-list.
LastCommandList stores an iterator pointing to the last command-list used.

You always submit a barrier after each command (for in-order semantics even within single command-list).
Why do you ever need to go back and submit more barriers?

Reply:
I've already explained this offline, here's the short answer:
I always submit a barrier "without event" after each command, and submit a barrier "with event" when command-list is closing. And also I submit a barrier "with event" in command-list1, when we have a switch between command-list1 to command-list2. (in that case, we need to go back because we can't predict it and it's done at runtime)